angular/projects/admin-nrpti/src/app/mines/mines-add-edit/mines-add-edit.component.ts

```diff
@@ -16,105 +17,197 @@ export class MinesAddEditComponent implements OnInit, OnDestroy {
-      this.isEditing = res.breadcrumb !== 'Add Mine';
-      if (this.isEditing) {
-        if (res && res.record && res.record[0] && res.record[0].data) {
-          this.record = new Mine(res.record[0].data);
-        } else {
-          alert('Error: could not load edit mine.');
-          this.router.navigate(['mines']);
-        }
+      if (res && res.mine && res.mine[0] && res.mine[0].data) {
+        this.mine = new Mine(res.mine[0].data);
       } else {
-        // Add
-        if (res && res.record) {
-          this.record = new Mine(res.record);
-        }
+        alert('Error: could not load edit mine.');
+        this.router.navigate(['mines']);
       }
+
       this.buildForm();
+
       this.loading = false;
+      this.loadingScreenService.setLoadingState(false, 'main');
+
       this._changeDetectionRef.detectChanges();
     });
   }
 
-  private buildForm() {
+  /**
+   * Derive static text strings.
+   *
+   * @memberof MinesAddEditComponent
+   */
+  populateTextFields() {
+    if (this.mine.dateUpdated) {
+      this.lastEditedSubText = `Last Edited on ${this.utils.convertJSDateToString(new Date(this.mine.dateUpdated))}`;
+    } else {
+      this.lastEditedSubText = `Added on ${this.utils.convertJSDateToString(new Date(this.mine.dateAdded))}`;
+    }
+  }
+
+  /**
+   * Build the formcontrols.
+   *
+   * @memberof MinesAddEditComponent
+   */
+  buildForm() {
     this.myForm = new FormGroup({
-      description: new FormControl((this.record && this.record.description) || ''),
-      summary:     new FormControl((this.record && this.record.summary) || ''),
-      type:        new FormControl((this.record && this.record.type) || ''),
-      links:       new FormControl((this.record && this.record.links.join()) || '')
+      description: new FormControl((this.mine && this.mine.description) || ''),
+      summary: new FormControl((this.mine && this.mine.summary) || ''),
+      type: new FormControl((this.mine && this.mine.type) || ''),
+      links: new FormArray(this.getLinksFormGroups()),
+      publish: new FormControl((this.mine && this.mine.read.includes('public')) || false)
+    });
+  }
+
+  /**
+   * Builds an array of links FormGroups, each with its own set of FormControls.
+   *
+   * @returns {FormGroup[]} array of links FormGroup elements
+   * @memberof MinesAddEditComponent
+   */
+  getLinksFormGroups(): FormGroup[] {
+    if (!this.mine || !this.mine.links || !this.mine.links.length) {
+      return [];
+    }
+
+    const links: FormGroup[] = [];
+
+    this.mine.links.forEach((link: Link) => {
+      links.push(
+        new FormGroup({
+          title: new FormControl(link.title || ''),
+          url: new FormControl(link.url || '')
+        })
+      );
     });
+
+    return links;
   }
 
+  /**
+   * Parses an array of links FormGroups into objects expected by the API.
+   *
+   * @returns {object[]} array of links objects
```

Comment:
Can this use the same `Link` type from line 96?

Reply:
yeah we should, will change